### PR TITLE
fixing error handler to stop panic

### PR DESCRIPTION
**Pull request by hqjang-pepper**

### CODE REVIEW

File: Shared/ThresholdKeyView.swift [hunk 1]
```diff
@@ -57,6 +57,7 @@ struct ThresholdKeyView: View {
                                 // try? await KeychainInterface.syncShare(threshold_key: threshold_key, share_index: nil)
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:57]: Where is the save and fetch logic from #4.

File: Shared/ThresholdKeyView.swift [hunk 2]
```diff
@@ -65,7 +66,10 @@ struct ThresholdKeyView: View {
                                 }
                                 
                                 // TODO: Add proper messages, 1/2 and 2/2 (old accounts, new accounts)
-                                alertContent = "\(totalShares) shares created"
+                                alertContent = "\(totalShares - required_share)/\(totalShares) shares created."
+                                if totalShares == 2 && required_share == 1 {
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:70]: No, comparison is incorrect, if total_shares < required_shares. 

This doesn't take into account the device share being added.

File: Shared/ThresholdKeyView.swift [hunk 3]
```diff
@@ -83,7 +87,7 @@ struct ThresholdKeyView: View {
                                 // Why does the application state Panic ??
                                 let key = try? await threshold_key.reconstruct()
                                 if key == nil {
-                                    alertContent = "Reconstruction failed"
+                                    alertContent = "Reconstruction failed. This is most likely caused by a non-zero value for required_share. "
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:90]: This message is incorrect and won't be reached since the try? in the above will fail.

File: Shared/ThresholdKeyView.swift [hunk 4]
```diff
@@ -65,7 +66,10 @@ struct ThresholdKeyView: View {
                                 }
                                 
                                 // TODO: Add proper messages, 1/2 and 2/2 (old accounts, new accounts)
-                                alertContent = "\(totalShares) shares created"
+                                alertContent = "\(totalShares - required_share)/\(totalShares) shares created."
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:69]: This message is incorrect.

If you have two shares and one is required, you do not have "1/2 shares created".

File: Shared/ThresholdKeyView.swift [hunk 5]
```diff
@@ -216,15 +235,21 @@ struct ThresholdKeyView: View {
...
                             Task {
-                                let question = "what's your password?"
-                                let answer = "blublublu"
-                                _ = try! await SecurityQuestionModule.change_question_and_answer(threshold_key: threshold_key, questions: question, answer: answer)
-                                let key_details = try! threshold_key.get_key_details()
-                                totalShares = Int(key_details.total_shares)
-                                threshold = Int(key_details.threshold)
+                                    let question = "what's your password?"
+                                    let answer = "blublublu"
+                                    let result = try! await SecurityQuestionModule.change_question_and_answer(threshold_key: threshold_key, questions: question, answer: answer)
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:240]: This can fail and others not part of this PR

File: Shared/ThresholdKeyView.swift (hunk 2 above)
Comment (hqjang-pepper) [reply Shared/ThresholdKeyView.swift]: fixed

File: Shared/ThresholdKeyView.swift (hunk 3 above)
Comment (hqjang-pepper) [reply Shared/ThresholdKeyView.swift:92]: fixed message

File: Shared/ThresholdKeyView.swift (hunk 4 above)
Comment (hqjang-pepper) [reply Shared/ThresholdKeyView.swift]: editied contition and message

File: Shared/ThresholdKeyView.swift (hunk 5 above)
Comment (hqjang-pepper) [reply Shared/ThresholdKeyView.swift]: my bad. Checked all buttons again and catched errors

File: Shared/ThresholdKeyView.swift (hunk 1 above)
Comment (hqjang-pepper) [reply Shared/ThresholdKeyView.swift]: I'll apply this on another PR

File: Shared/ThresholdKeyView.swift [hunk 6]
```diff
@@ -230,15 +236,22 @@ struct ThresholdKeyView: View {
...
-                                let index = shares.hex
 
-                                let key_details = try! threshold_key.get_key_details()
-                                totalShares = Int(key_details.total_shares)
-                                threshold = Int(key_details.threshold)
-                                shareIndexCreated = index
-                                alertContent = "You have \(totalShares) shares. New share with index, \(index) was created"
-                                showAlert = true
+                                let shares = try? await threshold_key.generate_new_share()
+                                if shares == nil {
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:241]:
```suggestion
do {
                                let shares = try await threshold_key.generate_new_share()
                                ...
catch {
...
}
```

File: Shared/ThresholdKeyView.swift [hunk 7]
```diff
@@ -230,15 +236,22 @@ struct ThresholdKeyView: View {
...
+                                    alertContent = "generate new share failed"
+                                    showAlert = true
+                                } else {
+                                    let index = shares!.hex
+                                    let key_details = try! threshold_key.get_key_details()
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:246]: These can crash still.

File: Shared/ThresholdKeyView.swift [hunk 8]
```diff
@@ -211,11 +212,16 @@ struct ThresholdKeyView: View {
...
-                                alertContent = "There are \(totalShares) available shares. \(key_details.required_shares) are required to reconstruct the private key"
-                                showAlert = true
+                                let key_details = try? threshold_key.get_key_details()
+                                if key_details == nil {
+                                    alertContent = "get key details failed"
+                                    showAlert = true
+                                } else{
+                                    totalShares = Int(key_details!.total_shares)
+                                    threshold = Int(key_details!.threshold)
+                                        alertContent = "You have \(totalShares) shares. \(key_details!.required_shares) are required to reconstruct the final key\n threshold : \(threshold)\n pub_key: (\(try! key_details!.pub_key.getX()) , \(try! key_details!.pub_key.getY())) \n share_descriptions: \(key_details!.share_descriptions)"
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:222]: Keep the previous message here.

File: Shared/ThresholdKeyView.swift [hunk 9]
```diff
@@ -468,25 +517,32 @@ struct ThresholdKeyView: View {
...
+                                    let shareOut = try threshold_key.output_share(shareIndex: index, shareType: nil)
 
-                                let result = try! ShareSerializationModule.serialize_share(threshold_key: threshold_key, share: shareOut, format: nil)
-                                alertContent = "serialize result is \(result)"
-                                showAlert = true
+                                    let result = try ShareSerializationModule.serialize_share(threshold_key: threshold_key, share: shareOut, format: nil)
+                                    alertContent = "serialize result is \(result)"
+                                    showAlert = true
+                                } catch {
+                                    // Handle the error here
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:535]:
```suggestion
```

File: Shared/ThresholdKeyView.swift [hunk 10]
```diff
@@ -468,25 +517,32 @@ struct ThresholdKeyView: View {
...
+                                    threshold = Int(key_details.threshold)
+                                    shareIndexCreated = index
 
-                                let shareOut = try! threshold_key.output_share(shareIndex: index, shareType: nil)
+                                    let shareOut = try threshold_key.output_share(shareIndex: index, shareType: nil)
 
-                                let result = try! ShareSerializationModule.serialize_share(threshold_key: threshold_key, share: shareOut, format: nil)
-                                alertContent = "serialize result is \(result)"
-                                showAlert = true
+                                    let result = try ShareSerializationModule.serialize_share(threshold_key: threshold_key, share: shareOut, format: nil)
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:531]: Hmm, this would be the same value I think.

With the absence of shareType and format it should produce hex in both cases.

There methods could also be changes to have the two above parameters have default values.

**Review by metalurgical — CHANGES_REQUESTED** [review]
Please see comments

File: Shared/ThresholdKeyView.swift (hunk 10 above)
Comment (hqjang-pepper) [reply Shared/ThresholdKeyView.swift:582]: Let's handle this at the another PR

File: Shared/ThresholdKeyView.swift [hunk 11]
```diff
@@ -243,11 +244,17 @@ struct ThresholdKeyView: View {
...
-                                let key_details = try! threshold_key.get_key_details()
-                                totalShares = Int(key_details.total_shares)
-                                threshold = Int(key_details.threshold)
-                                alertContent = "There are \(totalShares) available shares. \(key_details.required_shares) are required to reconstruct the private key"
-                                showAlert = true
+                                do {
+                                    let key_details = try threshold_key.get_key_details()
+                                    totalShares = Int(key_details.total_shares)
+                                    threshold = Int(key_details.threshold)
+                                        alertContent = "You have \(totalShares) shares. \(key_details.required_shares) are required to reconstruct the final key\n threshold : \(threshold)\n pub_key: (\(try key_details.pub_key.getX()) , \(try! key_details.pub_key.getY())) \n share_descriptions: \(key_details.share_descriptions)"
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:251]: The embedded `try!` can fail here.

Think I said to rather take the previous comment here which does not have this problem:
```suggestion
alertContent = "There are \(totalShares) available shares. \(key_details.required_shares) are required to reconstruct the private key"
                                showAlert = true
```

File: Shared/ThresholdKeyView.swift [hunk 12]
```diff
@@ -488,11 +518,16 @@ struct ThresholdKeyView: View {
...
                             Task {
-                                let seedResult = try!
-                                    SeedPhraseModule
-                                    .get_seed_phrases(threshold_key: threshold_key)
-                                print("result", seedResult)
-                                alertContent = "seed phrase is `\(seedResult[0].seedPhrase)`"
+                                do {
+                                    let seedResult = try SeedPhraseModule.get_seed_phrases(threshold_key: threshold_key)
+                                    if seedResult.isEmpty {
+                                        alertContent = "get seed phrase failed"
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:524]:
```suggestion
                                        alertContent = "No seed phrases set"
```

File: Shared/ThresholdKeyView.swift (hunk 10 above)
Comment (metalurgical) [reply Shared/ThresholdKeyView.swift:582]: Ok, then you wouldn't need to call serialize_share here for this PR.

**Review by metalurgical — CHANGES_REQUESTED** [review]
Much better, still a few things to resolve though.

File: Shared/ThresholdKeyView.swift (hunk 10 above)
Comment (hqjang-pepper) [reply Shared/ThresholdKeyView.swift:583]: what I changed in this PR was just changing try! to try.
no additional serialize_share call.

File: Shared/ThresholdKeyView.swift [hunk 13]
```diff
@@ -531,25 +569,31 @@ struct ThresholdKeyView: View {
                         Spacer()
                         Button(action: {
                             Task {
-                                let shareStore = try! await threshold_key.generate_new_share()
-                                let index = shareStore.hex
+                                do {
+                                    let shareStore = try await threshold_key.generate_new_share()
```
Comment (metalurgical) [inline Shared/ThresholdKeyView.swift:573]: Are we intentionally discarding the newly generated share, or should we be saving it?

Comment (metalurgical) [reply Shared/ThresholdKeyView.swift:573]: Hmm, this one can left for another PR.

**Review by metalurgical — APPROVED** [review]
lgtm
